admin/modules/style/themes.php

```diff
@@ -303,162 +396,282 @@
+					} else {
+						// Save them to a file in a temporary directory
+						$tmp_dir = create_temp_dir();
+						$zip_filepath = $tmp_dir.'/theme.zip';
+						if (!$tmp_dir || !file_put_contents($zip_filepath, $contents)) {
+							$errors[] = $lang->error_failed_to_save_remote_file;
+						}
 					}
-					// Get the contents
-					$contents = @file_get_contents($_FILES['local_file']['tmp_name']);
-					// Delete the temporary file if possible
-					@unlink($_FILES['local_file']['tmp_name']);
-					// Are there contents?
-					if(!trim($contents))
-					{
-						$errors[] = $lang->error_uploadfailed_nocontents;
+					break;
+			}
+		}
+
+		if (!$errors) {
+			assert(!empty($zip_filepath));
+			if ($tmp_dir) {
+				$extract_dir = $tmp_dir.'/extracted';
+			} else	$tmp_dir = $extract_dir = create_temp_dir();
+			if (!$extract_dir) {
+				$errors[] = $lang->error_failed_to_create_tmpdir;
+			}
+
+			if (!$errors) {
+				// Unzip the file into the temporary directory.
+				if (!class_exists('ZipArchive')) {
+					$errors[] = $lang->error_no_ziparchive_for_theme;
+				} else {
+					$za = new ZipArchive;
+					$res = $za->open($zip_filepath);
+					if ($res !== true) {
+						$errors[] = $lang->sprintf($lang->error_theme_unzip_open_failed, $res);
+					} else {
+						if (!$za->extractTo($extract_dir)) {
+							$errors[] = $lang->error_theme_unzip_failed;
+						} else {
+							$top_lvl_files = glob("$extract_dir/*");
+							if (count($top_lvl_files) != 1) {
+								rmdir_recursive($tmpdir);
+								$errors[] = $lang->error_theme_unzip_multi_or_none_root;
+							} else {
+								$codename = basename($top_lvl_files[0]);
+								if (!is_valid_theme_code($codename)) {
+									$errors[] = $lang->sprintf($lang->error_invalid_theme_archive_tld, htmlspecialchars_uni($codename));
+								} else {
+									$unzipped_path = "$extract_dir/$codename";
+								}
+							}
+						}
+						$za->close();
 					}
 				}
 			}
-			else if(!empty($mybb->input['url']))
-			{
-				// Get the contents
-				$contents = @fetch_remote_file($mybb->input['url']);
-				if(!$contents)
-				{
-					$errors[] = $lang->error_local_file;
+
+			@unlink($zip_filepath);
+		}
+
+		if (!$errors) {
+			$infofile2 = '';
+			$infofile = "$unzipped_path/theme.json";
+			if (!($themeinfo = read_json_file($infofile))) {
+				$errors[] = $lang->sprintf($lang->error_bad_unzipped_json_file, htmlspecialchars_uni($infofile));
+			} else if (empty($themeinfo['codename'])) {
+				$errors[] = $lang->sprintf($lang->error_missing_theme_file_property, 'codename', htmlspecialchars_uni($infofile));
+			} else if ($themeinfo['codename'] != $codename) {
+				$errors[] = $lang->sprintf($lang->error_codename_mismatch, htmlspecialchars_uni($themeinfo['codename']), htmlspecialchars_uni($codename));
+			} else if (file_exists(MYBB_ROOT."inc/themes/{$codename}")) {
+				$infofile2 = MYBB_ROOT."inc/themes/{$codename}/theme.json";
```

Comment:
is it in general a good idea to have themes inside the `inc` folder which usually is protected from public access? Or are the theme assets at a different location? If so, I would suggest to keep them in one location and rather force their directory structure to have a `public` folder with assets that must be accessible from outside while the rest can be protected via .htaccess or nginx rules.